////////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2014-2023, Lawrence Livermore National Security, LLC.
// Produced at the Lawrence Livermore National Laboratory.
// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
//
// LLNL-CODE-697807.
// All rights reserved.
//
// This file is part of LBANN: Livermore Big Artificial Neural Network
// Toolkit. For details, see http://software.llnl.gov/LBANN or
// https://github.com/LLNL/LBANN.
//
// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
// may not use this file except in compliance with the License.  You may
// obtain a copy of the License at:
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the license.
////////////////////////////////////////////////////////////////////////////////

syntax = "proto3";

import "datatype.proto";
import "optimizers.proto";

package lbann_data;

message Weights {
  string name = 1;
  Optimizer optimizer = 2;
  Initializer initializer = 3;
  DataType datatype = 4;
}

message Initializer {
  oneof initializer_type {
    ConstantInitializer constant_initializer = 20;
    ValueInitializer value_initializer = 21;
    NumpyInitializer numpy_initializer = 22;
    UniformInitializer uniform_initializer = 23;
    NormalInitializer normal_initializer = 24;
    GlorotNormalInitializer glorot_normal_initializer = 25;
    GlorotUniformInitializer glorot_uniform_initializer = 26;
    HeNormalInitializer he_normal_initializer = 27;
    HeUniformInitializer he_uniform_initializer = 28;
    LeCunNormalInitializer lecun_normal_initializer = 29;
    LeCunUniformInitializer lecun_uniform_initializer = 30;
  }

  // Weight initializers

  /** @brief Fill weights with a single constant value. */
  message ConstantInitializer {
    double value = 1;
  }
  /** @brief Fill weights with values from a list.
   *
   *  The number of weight entries must exactly match the number of
   *  provided values. Choose one of ``values`` and ``values_d``.
   *  If double-precision weights are necessary (e.g., for debugging), use
   *  ``values_d``.
   *
   *  @note Most weights are stored in row-major order. However, the
   *  fully-connected layer's linearity weights are column-major.
   */
  message ValueInitializer {
    repeated float values = 1 [packed = true];
    repeated double values_d = 2 [packed = true];
  }
  /** @brief Fill weights with values from a NumPy file.
   *
   *  Expects a .npy file with float32 or float64 values in C-style,
   *  row-major order.
   */
  message NumpyInitializer {
    /// NumPy file
    string file = 1;
  }
  message UniformInitializer {
    double min = 1;
    double max = 2;
  }
  message NormalInitializer {
    double mean = 1;
    double standard_deviation = 2;
  }
  message GlorotNormalInitializer {}
  message GlorotUniformInitializer {}
  message HeNormalInitializer {}
  message HeUniformInitializer {}
  message LeCunNormalInitializer {}
  message LeCunUniformInitializer {}
}
